hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java

```diff
@@ -47,22 +47,28 @@ public TemplateRenderer(String template) {
 		while ( i < len ) {
 			char c = template.charAt( i );
 			if ( c == '?' ) {
-				chunkList.add( chunk.toString() );
-				chunk.delete( 0, chunk.length() );
-
 				while ( ++i < template.length() ) {
 					c = template.charAt( i );
 					if ( Character.isDigit( c ) ) {
+						if ( chunk.length() > 0 ) {
+							chunkList.add( chunk.toString() );
+							chunk.setLength( 0 );
+						}
 						index.append( c );
 					}
 					else {
+						if ( index.length() == 0 ) {
+							chunk.append( '?' );
+						}
 						chunk.append( c );
 						break;
 					}
 				}
 
-				paramList.add( Integer.valueOf( index.toString() ) );
-				index.delete( 0, index.length() );
+				if ( index.length() > 0 ) {
+					paramList.add( Integer.valueOf( index.toString() ) );
+					index.setLength( 0 );
+				}
```

Comment:
The basic idea is simple. If '?' is not followed by digit(s), it should be teated as normal chunk char. However, the code seems unnecessarily complicated and we might refactor the logic by simple RE processing instead.